### DISCUSSION TITLE
Why can't I get query params in getStaticProps?

### DISCUSSION THREAD

**Discussion by alioshr** [vercel/next.js#17529] [Help] [answered]
I will make it simple:

Link: ```http://localhost:3000/product/bioCloths?activeCategory=medium&id=0```

File path: ```pages/product/[product].js.```

Expected: ``` product: "bioCloths, activeCategory: "medium", id: "0"```

using ```getStaticProps``` at ```[product].js```:

```
const ProductDetails = (props) => {
  console.log("initial props", props);

  return <div>product details...</div>
};

export default ProductDetails;

export async function getStaticProps(context) {

  return {
    props: {
     context: context
    },
  };
}

export async function getStaticPaths() {
  return {
    paths: [{ params: { product: "1" } }, { params: { product: "2" } }],
    fallback: true,
  };
}
```
```Props``` returns: ```context: params: product: "bioCloths"```, excluding the query params.

Now if I use the deprecated ```getInitialProps``` instead:

```
ProductDetails.getInitialProps = (context) => {
  const activeCategory = context.query.activeCategory;
  const id = context.query.id;
  const product = context.query.product;

  return {
    activeCategory: activeCategory,
    id: id,
    product: product,
  };
};
```
```props``` logs ```activeCategory: "medium" id: "0" product: "bioCloths"```

I need to get these all of these ```props``` so I can ```fetch``` ```data``` before the client mounts.

I understand that ```getInitialProps``` is now deprecated, but it works. Why is not ```getStaticProps``` working, and should I use it instead of ```serverSideProps```?

This is an e-commerce, and there is no way I can set ```getStaticPaths``` for hundreds of possibilities, to work along with ```getStaticProps``` so I guess that in this case I should use SSR ```getInitialProps``` or ```getServerSideProps```, correct?

P.S - ```getServerSideProps``` hits me an error stating that it will only accept ```.json``` files.

**Comment by lfades** [comment, marked as answer]
`getStaticProps` generates the page at build time. There's no possible way to know the custom query params your visitors can use at build time.

`getInitialProps` and `getServerSideProps` can know all the possible query params because the page is generated at runtime, whenever you receive a new request.

You can learn more about the above here: https://nextjs.org/docs/basic-features/pages#two-forms-of-pre-rendering

**Reply by alioshr** [reply]
also, if I use a mask ```<Link href="href..." as="mask...">```, I am not getting ```getInitialProps```. Should I parse the URL myself in this case?

**Reply by lucasbasquerotto** [reply]
If I understand correctly, `getStaticProps()` generates different HTMLs for different slugs, so when a user acesses a url that has `slug1` he will see `page1` and with `slug2` he would see `page2`.

Is it not possible (at least as a feature request) to define query parameters to generate different pages, even if not in the path?

Example for `mypage/:slug`:

```
mypage/slug1 -> page1
mypage/slug2 -> page2
mypage/slug1?query=value1 -> page3
mypage/slug1?query=value2 -> page4
```

(they would behave similarly to `mypage/[slug]/[...rest]`, although `rest` in this case would not be a part of the path.

In this case, I would expect to generate the pages statically, and returning the statically generated page based on the slug and query parameters.

This would be very important if you want static pages with pagination that have the page number in the query, or for pages that have a search mechanism and the search params are sent in the query (which is natural, although it could be passed in the path too).

The query parameters that would be considered could be specified in a method similar to `getStaticProps()` and `getServerSideProps()`, but to define query parameters to be considered to render different static pages. Like the following proposal:

```js
export function getStaticParams() {
    return ['page', 'order'];
}
```

(this way I would also able to ignore parameters like Facebook `fbid` that are not relevant to render the page) 

In this case it would generate different static pages for `/example`, `/example?page=2` and `/example?page=2&order=desc`

Related: https://github.com/vercel/next.js/discussions/17616

**Reply by edoardo-bluframe** [reply]
That isn't exactly true right?

You cannot predict `slug1` and `slug` in a path eitther. And that is why we have `getStaticPaths`

Can we not have a `getStaticPaths` that also supports a set of predefined query params?

The same way it already does with paths?

**Comment by lucasbasquerotto** [comment]
Related: https://github.com/vercel/next.js/discussions/17269
